### PR TITLE
Restore Labels, Points and Shapes layer mode after hold key

### PR DESCRIPTION
**Pull request by Czaki**
# Description

In this PR I implement a mechanism for restoring attribute value if the keyboard key is held longer. It is used for `Labels.mode` related actions. 

If the key is pressed, then the mode is changed, but if the key is kept above 0.5 seconds (could be changed in settings - accessibility requirements), then the value is restored on key release. 

The general idea of this PR is that bind action could return a function, that will be called if between the press, and release of the key is above 0.5 seconds. 

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

Alternative and extension of #4696

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change
- [ ] example: I check if my changes works with both PySide and PyQt backends
      as there are small differences between the two Qt bindings.  

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

### CODE REVIEW

File: napari/layers/utils/_tests/test_layer_utils.py
Comment (jni) [inline napari/layers/utils/_tests/test_layer_utils.py:515]: Wow. ❤️

File: napari/layers/utils/layer_utils.py
Comment (jni) [inline napari/layers/utils/layer_utils.py]: @Czaki what is this meant to cover? (Since codecov complains that tests don't cover it, it's unclear.)

"name" is also unclear as a variable name. I see below that this is probably `first_parameter_name`? Why is only the first parameter name important here? I think some assumptions of the function should be spelled out in the docstring.

File: napari/settings/_application.py
Comment (jni) [inline napari/settings/_application.py]: A longer description would also be good, for example: "This affects certain actions where a short press and a long press have different behaviors, such as changing the mode of a layer permanently or only during the long press."

File: napari/utils/key_bindings.py
Comment (jni) [inline napari/utils/key_bindings.py]: The logic here is now a bit more complicated — `gen()` in this case is not a generator, right? So it's probably not the right name? Maybe it is worth adding a comment explaining the two paths. "If it's a tuple, it is a function that will reset the mode back together with a timestamp. Otherwise, it is a generator that..." (and tbh I don't understand the logic at that point 😅)

**Review by jni — COMMENTED** [review]
Fantastic, thank you @Czaki! I only have very minor code comments/requests for clarification.

Do you think this would supersede #4696?

File: napari/settings/_application.py
Comment (Czaki) [reply napari/settings/_application.py]: Thanks. I have no idea for a longer description, so I copy this one.

File: napari/layers/utils/layer_utils.py
Comment (Czaki) [reply napari/layers/utils/layer_utils.py]: update to exception instead of silent fail.

**Comment by Czaki** [comment]
@jni This PR is a little different from #4696, but it removes the need to have a separate shortcut. So we may treat this as supersede of this PR. 

There is one thing left:
https://github.com/napari/napari/blob/458a7e0f0a15c17fab12cb5ae45cbfc591fa303e/napari/layers/labels/labels.py#L38-L51

all such strings should be updated to reflect current settings. Especially cannot be global strings. But I'm unsure if it should be done here or in separate PR.

File: napari/utils/key_bindings.py
Comment (Czaki) [reply napari/utils/key_bindings.py]: Could you check now?

File: napari/layers/shapes/shapes.py
Comment (Czaki) [inline napari/layers/shapes/shapes.py]: I think that only a subset of this actions should be selected

File: napari/layers/points/points.py
Comment (jni) [inline napari/layers/points/points.py]: Can you add a docstring to these functions? At first I didn't understand what they were doing — just making the help string, right? I might suggest `_mode_help_string`?

I might also suggest something like the following:

```
"quick press <{shortcut}> to switch to {mode} ({desc}), hold and release <{shortcut}> to temporarily switch while holding"
```

File: napari/layers/shapes/shapes.py
Comment (jni) [reply napari/layers/shapes/shapes.py]: Which subset and why?

Comment (jni) [reply napari/layers/shapes/shapes.py]: (Also, rather than get stuck in a long discussion about this, I suggest removing Shapes layer support in this PR and then adding it in an followup PR.)

Comment (Czaki) [reply napari/layers/shapes/shapes.py]: The produces string is very long an maybe it is worth to not inform about every shortcuts for Shapes layer.

File: napari/layers/points/points.py
Comment (Czaki) [reply napari/layers/points/points.py]: > `"quick press <{shortcut}> to switch to {mode} ({desc}), hold and release <{shortcut}> to temporarily switch while holding"`

This will be to long. Such text will occupy the whole bottom for one shortcut bar without a place for information about other shortcuts. 
I see that it may increase discoverability for the new mode, but for me, it should instead help to remind about currently set shortcuts, and information about long hold should be distributed another way.

Comment (jni) [reply napari/layers/points/points.py]: oh I thought this text was used for a tooltip?

Comment (Czaki) [reply napari/layers/points/points.py]: No. It is used on statusbar. 

Zrzut ekranu z 2022-07-23 16-11-38

Here docstring from base `Layer` (but not copied to inheriting class)

```
    help : str
        Displayed in status bar bottom right.
```

Comment (jni) [reply napari/layers/points/points.py]: oh yeah that's crazy, and they all get concatenated together. Maybe we should rethink that anyway... But ok how it is for now. 🙏

**Review by jni — APPROVED** [review]

File: napari/layers/points/points.py
Comment (Czaki) [reply napari/layers/points/points.py]: Should I open a separate issue to not lose this information?

**Comment by jni** [comment]
@Czaki what do you think about merging this and refining messages + command subsets later?

@napari/core-devs does anyone have any objections to merging this ~soon?

**Comment by Czaki** [comment]
> @Czaki what do you think about merging this and refining messages + command subsets later?

Agree. Just waiting if someone else has an opinion. But there are 2 days since your approval, so if today there will be no comment that requires some change then merge.

File: napari/layers/labels/labels.py
Comment (brisvag) [inline napari/layers/labels/labels.py:716]: Is there a reason for this change? Just curious, cause I don't see what it would change.

File: napari/layers/points/_points_key_bindings.py
Comment (brisvag) [inline napari/layers/points/_points_key_bindings.py]: Uhm, there seems to be a copy-pase issue here, it should probably be `points`, noy `label`. (And same for all the following)

File: napari/layers/points/points.py
Comment (brisvag) [inline napari/layers/points/points.py]:
```suggestion
    name : str
```

File: napari/layers/shapes/_shapes_key_bindings.py
Comment (brisvag) [inline napari/layers/shapes/_shapes_key_bindings.py]: Same as above.

**Review by brisvag — CHANGES_REQUESTED** [review]
I'm a bit confused by the code, so I'll defer to others for the details. If I understand correctly, there are a few things that are misnamed.

File: napari/layers/labels/labels.py
Comment (Czaki) [reply napari/layers/labels/labels.py:716]: change from a tuple to set. It is a little more pythonic.

File: napari/layers/points/_points_key_bindings.py
Comment (Czaki) [reply napari/layers/points/_points_key_bindings.py]: it should be `layer`, not `label` in every place

File: napari/layers/labels/labels.py
Comment (brisvag) [reply napari/layers/labels/labels.py:716]: Why? Is it faster? Or just more "correct"? I feel like I've only ever seen the tuple version :)

File: napari/layers/points/_points_key_bindings.py
Comment (brisvag) [inline napari/layers/points/_points_key_bindings.py]: You still call the new decorator `label`; same for all the other layer types.

File: napari/layers/labels/labels.py
Comment (jni) [reply napari/layers/labels/labels.py:716]: In theory it's faster, O(1), to check whether something is in a set vs a tuple (O(N)). For a 2-tuple? Probably about the same. I would agree with @Czaki that the "preferred" form for `if x in xs` is for xs to be a set, unless it is *already* a list or a tuple for some prior reason.

File: napari/layers/points/_points_key_bindings.py
Comment (Czaki) [reply napari/layers/points/_points_key_bindings.py]: I fixed the name of the function called, but not the name of the defined function..

File: napari/layers/labels/labels.py
Comment (jni) [reply napari/layers/labels/labels.py:716]: Here you go:

```
In [184]: %timeit 1 in {1, 2}
83.2 ns ± 3.74 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [185]: %timeit 1 in (1, 2)
70.4 ns ± 3.14 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [186]: %timeit 1 in (2, 1)
104 ns ± 8.75 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [187]: %timeit 1 in {2, 1}
83.6 ns ± 2.71 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

Comment (jni) [reply napari/layers/labels/labels.py:716]: So the advantage is measureable even for two elements. The only time tuple is faster is if you hit the very first element. tbh I'm surprised by this and it points to some serious optimisation of sets/dicts in Python!

Comment (Czaki) [reply napari/layers/labels/labels.py:716]: Just more correct. With a set of sizes, two the speeds should be the same. For the larger one, it may have an impact. 

I change it because of the linter highlight.

Comment (tlambert03) [reply napari/layers/labels/labels.py:716]: also
```python
In [1]: _set = set(range(1000))

In [2]: _tup = tuple(range(1000))

In [3]: %timeit 100 in _set
27.9 ns ± 0.0384 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [4]: %timeit 100 in _tup
632 ns ± 2.39 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

(edit: sorry... stupid addition probably given that you specifically mentioned 2-tuples.  ignore me :joy:)

Comment (brisvag) [reply napari/layers/labels/labels.py:716]: Wow, I never considered that!

Worth noting that constructing a set seems to be slower:
```py
In [1]: %timeit set(range(1000))
21.2 µs ± 2.07 µs per loop (mean ± std. dev. of 7 runs, 10,000 loops each)

In [2]: %timeit tuple(range(1000))
12.9 µs ± 155 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

So unless the set exists already, this time should also be accounted for (which @jni already did for the 2-tuple case).

```py
In [10]: %timeit 10 in set(range(10))
362 ns ± 8.62 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [11]: %timeit 10 in tuple(range(10))
325 ns ± 9.09 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

So in practice, a tuple might be better for this kind of case :P

Comment (Czaki) [reply napari/layers/labels/labels.py:716]: Yes. But

```python
In [1]: %timeit 10 in {0,1,2,3,4,5,6,7,8,9}
22.4 ns ± 0.918 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [2]: %timeit 10 in (0,1,2,3,4,5,6,7,8,9)
111 ns ± 0.54 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
```

As build set from the constructor and using literal constructor is different.

Comment (brisvag) [reply napari/layers/labels/labels.py:716]: Damn, crazy!

File: napari/layers/shapes/_shapes_key_bindings.py
Comment (brisvag) [inline napari/layers/shapes/_shapes_key_bindings.py]: For consistency, we should stick with `shapes`, rather than `shape`. Sorry for the nitpicking on the names @Czaki :P

Comment (Czaki) [reply napari/layers/shapes/_shapes_key_bindings.py]: You found my weakpoint.

File: napari/layers/labels/labels.py
Comment (Czaki) [reply napari/layers/labels/labels.py:716]: Some longer comparisons:
```python
In [1]: %timeit 10 in {0,1,2,3,4,5,6,7,8,9}
22.1 ns ± 0.0787 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [2]: %timeit 10 in (0,1,2,3,4,5,6,7,8,9)
114 ns ± 3.87 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)

In [3]: %timeit 10 in {i for i in range(10)}
537 ns ± 9.94 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [4]: %timeit 10 in [i for i in range(10)]
543 ns ± 7.96 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [5]: %timeit 10 in set(range(10))
334 ns ± 1.06 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [6]: %timeit 10 in list(range(10))
357 ns ± 5.39 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)

In [7]: %timeit 10 in tuple(range(10))
320 ns ± 3.06 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

where `set(range(10))` is faster than `{i for i in range(10)}`

**Review by brisvag — APPROVED** [review]
All good on my side!

File: napari/layers/labels/labels.py
Comment (sofroniewn) [inline napari/layers/labels/labels.py]: I don't love that we now have this import and the `get_settings` call in here. In general it moves our layer models in the direction of more coupled to the rest of the code base and more complicated - I'd much rather them be moving in the other direction. Can we avoid this pattern here?

**Review by sofroniewn — CHANGES_REQUESTED** [review]
Left one comment about code complexity/ coupling. Would be great if it could at least be commented on before merge

File: napari/layers/labels/labels.py
Comment (Czaki) [reply napari/layers/labels/labels.py]: we could move whole mode string generation outside layers implementation. Then based on the active layer and selected mode, prepare mode string.

Comment (sofroniewn) [reply napari/layers/labels/labels.py]: i think that would be much better - would that also allow us to reduce some of the duplicated code across layer types?

File: napari/layers/base/base.py
Comment (Czaki) [inline napari/layers/base/base.py:1011]: just fix type annotation

File: napari/layers/labels/_labels_key_bindings.py
Comment (Czaki) [inline napari/layers/labels/_labels_key_bindings.py:66]: such a list with string (name) and mode could be used to generate functions, but it may reduce code readability.

**Review by Czaki — COMMENTED** [review]
I move implementation to the `ViewerModel`, but maybe I should move it to `Viewer` or event Main Window. What is your opinion @sofroniewn

**Comment by sofroniewn** [comment]
> I move implementation to the ViewerModel, but maybe I should move it to Viewer or event Main Window. What is your opinion @sofroniewn

Looks like we already import get_settings in viewermodel. Maybe long term we remove need for it there, but this is much better. method is private so can be moved easily.

Thanks for making change!

**Review by sofroniewn — APPROVED** [review]

**Comment by jni** [comment]
Thanks for spotting that @sofroniewn, this is indeed an improvement! Since this has been open for some time and already has three approvals, I'm gonna go ahead and merge! 🎉

**Pull request merged**
